### PR TITLE
Modify URL related functions to use new structs

### LINKED ISSUE #1288: Modify URL to use new structs
- [x] Add missing fields
- [x] Update all functions that refer to URL struct to new structs (machine readable)

### PR DESCRIPTION
**Pull request by surajnarwade**
## What is the purpose of this change? What does it change?

Refactor `url` portion to use new structs throughtout and remove old url struct and all it's usage

## Was the change discussed in an issue?
fixes #1288

## How to test changes?

CI should be :green_heart:  and test all URL related commands

### CODE REVIEW

File: pkg/url/url.go [hunk 1]
```diff
@@ -77,20 +69,24 @@ func List(client *occlient.Client, componentName string, applicationName string)
...
-			URL:      r.Spec.Host,
-			Protocol: getProtocol(r),
-			Port:     r.Spec.Port.TargetPort.IntValue(),
-		})
+		a := getMachineReadableFormat(r)
+		urls = append(urls, a)
 	}
 
-	return urls, nil
+	urlList := UrlList{
```
Comment (girishramnani) [inline pkg/url/url.go]: we can use `NewURLList()` here to hide the manual setting of `TypeMeta` and `ListMeta`.

File: pkg/url/url.go [hunk 2]
```diff
@@ -102,8 +98,8 @@ func getProtocol(route routev1.Route) string {
 }
 
 // GetURLString returns a string representation of given url
-func GetURLString(url URL) string {
-	return url.Protocol + "://" + url.URL
+func GetURLString(protocol, URL string) string {
+	return protocol + "://" + URL
```
Comment (girishramnani) [inline pkg/url/url.go]: wouldn't the URL have the protocol @surajnarwade ?

File: pkg/url/url.go (hunk 1 above)
Comment (girishramnani) [reply pkg/url/url.go]: or `getMachineReadableFormat`

File: tests/e2e/json_test.go [hunk 3]
```diff
@@ -55,8 +55,13 @@ var _ = Describe("odojsonoutput", func() {
 		It("should be able to list url", func() {
 			actual := runCmdShouldPass("odo url list -o json")
 			url := runCmdShouldPass("oc get routes myurl-myapp -o jsonpath={.spec.host}")
-			desired := fmt.Sprintf(`{"kind":"List","apiVersion":"odo.openshift.io/v1alpha1","metadata":{},"items":[{"kind":"url","apiVersion":"odo.openshift.io/v1alpha1","metadata":{"name":"myurl","creationTimestamp":null},"spec":{"path":"%s","port":8080}}]}`, url)
+			desired := fmt.Sprintf(`{"kind":"List","apiVersion":"odo.openshift.io/v1alpha1","metadata":{},"items":[{"kind":"url","apiVersion":"odo.openshift.io/v1alpha1","metadata":{"name":"myurl","creationTimestamp":null},"spec":{"path":"%s","protocol":"http","port":8080}}]}`, url)
```
Comment (girishramnani) [inline tests/e2e/json_test.go]: going over the 80 character length, if we can make this into more than one lines

File: pkg/component/component.go [hunk 4]
```diff
@@ -38,7 +38,7 @@ type Description struct {
 	ComponentName      string                `json:"componentName,omitempty"`
 	ComponentImageType string                `json:"type,omitempty"`
 	Path               string                `json:"source,omitempty"`
-	URLs               []urlpkg.URL          `json:"url,omitempty"`
+	URLs               urlpkg.UrlList        `json:"url,omitempty"`
```
Comment (girishramnani) [inline pkg/component/component.go]: if this is a list shouldn't he json tag be `urlList` or `urls`?

File: pkg/url/url.go (hunk 2 above)
Comment (surajnarwade) [reply pkg/url/url.go]: I didn't get this ?

Comment (girishramnani) [reply pkg/url/url.go]: oh I mis-read the `URL` here as the struct `URL`

Comment (girishramnani) [reply pkg/url/url.go]: resolving

File: pkg/component/component.go (hunk 4 above)
Comment (surajnarwade) [reply pkg/component/component.go]: removed `tags`, not needed anymore

File: pkg/url/url.go (hunk 1 above)
Comment (surajnarwade) [reply pkg/url/url.go]: :+1:

**Comment by amitkrout** [comment]
@surajnarwade Can you please fill the pr description area

File: tests/e2e/json_test.go (hunk 3 above)
Comment (amitkrout) [reply tests/e2e/json_test.go]: Character length seems ```OK``` to me as ```golint``` does not report anything in https://github.com/redhat-developer/odo/pull/1330#issuecomment-463571158

File: tests/e2e/json_test.go [hunk 5]
```diff
@@ -55,8 +55,13 @@ var _ = Describe("odojsonoutput", func() {
...
 			areEqual, _ := compareJSON(desired, actual)
+			if !areEqual {
+
```
Comment (amitkrout) [inline tests/e2e/json_test.go]: @surajnarwade Any idea behind not to assert the mismatch ? Seems it prints the value rather verifying it.

File: pkg/component/component.go [hunk 6]
```diff
@@ -35,14 +35,14 @@ const componentNameMaxLen = -1
...
-	ComponentName      string                `json:"componentName,omitempty"`
-	ComponentImageType string                `json:"type,omitempty"`
-	Path               string                `json:"source,omitempty"`
-	URLs               []urlpkg.URL          `json:"url,omitempty"`
-	Env                []corev1.EnvVar       `json:"environment,omitempty"`
-	Storage            []storage.StorageInfo `json:"storage,omitempty"`
-	LinkedComponents   map[string][]string   `json:"linkedComponents,omitempty"`
-	LinkedServices     []string              `json:"linkedServices,omitempty"`
+	ComponentName      string
+	ComponentImageType string
```
Comment (girishramnani) [inline pkg/component/component.go]: why did we remove the json tags? wouldn't this default to json's `snake_case` instead of `CamelCase`?

Comment (surajnarwade) [reply pkg/component/component.go]: we don't need json tags in this struct, since we have machine readable struct now

Comment (girishramnani) [reply pkg/component/component.go]: the types.go have the tags necessary

Comment (girishramnani) [reply pkg/component/component.go]: resolving

**Comment by girishramnani** [comment]
@surajnarwade might wanna change the description of the PR to not close the issue, this PR only solves the 1 part of the issue. The second part of `adding missing fields` is still left.

**Comment by surajnarwade** [comment]
@girishramnani , current struct has all the fields required for URL, so I think we covered missing field point too

**Comment by girishramnani** [comment]
got it, so we can merge this once the CI goes green

**Comment by girishramnani** [comment]
Is the circleCI failure a flake @surajnarwade

**Comment by amitkrout** [comment]
> Is the circleCI failure a flake @surajnarwade

@girishramnani Yes, We have a issue [1] for that too
[1] https://github.com/redhat-developer/odo/issues/1322

**Pull request merged**